### ISSUE TITLE
Next.js server action not getting executed

### ISSUE THREAD

**Issue by RareSecond** [TanStack/query#7934]
### Describe the bug

Please bear with me, because I have no idea how to best start. I've also done my best to narrow this down as much as possible, but I haven't been able to provide a minimal reproducible example.

# Setting the stage

We were noticing more and more lately that our (Cypress) tests are flakey so I started to investigate it. I've managed to track it down to a custom hook not resolving with the correct data.

The hook, `useCurrentUser`, is in essence a very simple hook

```tsx
import { useQuery } from "@tanstack/react-query";

import { getCurrentUserWithRoles } from "@/actions/users";

export const useCurrentUser = () => {
  const { data: currentUserWithRoles } = useQuery({
    queryFn: async () => {
      return getCurrentUserWithRoles();
    },
    queryKey: ["currentUserWithRoles"],
  });

  if (!currentUserWithRoles) {
    return null;
  }

  return currentUserWithRoles;
};
```

`getCurrentUserWithRoles` is also a very simple server action that does some database fetching.

```ts
"use server";

import { AuthDAL } from "@/data/auth";

export async function getCurrentUserWithRoles() {
  return await AuthDAL.getCurrentUserWithRoles();
}
```

However, when running `cypress open` and running a test that relies on the `useCurrentUser` returning correct information, the hook **sometimes** fails to return any data.

# What I've tried

### Adding logs

If I add logs to the necessary places, I can see where it goes wrong

```tsx
    queryFn: async () => {
      console.log("About to get roles");
      const res = getCurrentUserWithRoles();
      console.log("res: ", res);

      return res;
    },
```

```ts
export async function getCurrentUserWithRoles() {
  console.log("Inside server action");
  return await AuthDAL.getCurrentUserWithRoles();
}
```

The first log gets printed, but the other two don't appear. So for some reason, it seems that it never reaches the server action.

### Removing React Query

If I remove React Query and just rely on `useEffect` and `useState`, there's no flakiness and test passes 100% of the time

```tsx
export const useCurrentUser = () => {
  const wasCalled = useRef(false);
  const [currentUserWithRoles, setCurrentUserWithRoles] = useState<Awaited<
    ReturnType<typeof getCurrentUserWithRoles>
  > | null>(null);

  useEffect(() => {
    if (!wasCalled.current) {
      wasCalled.current = true;

      getCurrentUserWithRoles().then((res) => {
        setCurrentUserWithRoles(res);
      });
    }
  }, []);

  if (!currentUserWithRoles) {
    return null;
  }

  return currentUserWithRoles;
};
```

### Replacing the server action with an actual API call

I also went ahead and created an actual API call (via `route.ts`) where I returned hardcoded data. And again, it works 100% of the time.

```ts
import { NextResponse } from "next/server";

export async function POST() {
  return NextResponse.json({
    groupRoles: [
      {
        groupIdId: "77fc6325-96ff-449b-91a8-cd0a8c79e019",
        role: "ADMIN",
      },
      {
        groupId: "4b102177-be72-4910-a46d-90060ca7b178",
        role: "ADMIN",
      },
    ],
    firstName: "Fred",
    id: "33d4cde9-c366-4c77-9669-decdf0e0c6ad",
    lastName: "Flinstone",
    appRole: "ADMIN",
  });
}
```

```tsx
export const useCurrentUser = () => {
  const { data: currentUserWithRoles } = useQuery({
    queryFn: async () => {
      // return getCurrentUserWithRoles();
      const call = await fetch("/api/v1/test", {
        method: "POST",
      });

      const res = await call.json();

      return res;
    },
    queryKey: ["currentUserWithRoles"],
  });

  if (!currentUserWithRoles) {
    return null;
  }

  return currentUserWithRoles;
};
```

### Replacing Cypress with Playwright

I even rewrote the test in Playwright but to no avail. The tests are still just as flaky, feeling even worse (but that may be my mental state acting up by now).

# Some extra insights

However, the switch to Playwright has brought some extra insights that may be helpful

- If I only enable one of the tree browsers in Playwright (so `chromium`, `firefox` and `webkit`), the tests always pass for `chromium` and `firefox`, and fail way less for `webkit` (10% of the times instead of 70%)
- Same scenario for setting `workers` to 1 in Playwright
- I tried [this suggestion](https://github.com/TanStack/query/issues/1657#issuecomment-1912378686) to no avail
- All libraries are at their latest versions

### Your minimal, reproducible example

I tried, but can't condense it enough to have it always fail

### Steps to reproduce

/

### Expected behavior

I expect React Query to actually fire the Next.js server action and return the data

### How often does this bug happen?

Often

### Platform

- OS: MacOS
- Browser: Multiple

### Tanstack Query adapter

react-query

### TanStack Query version

v5.52.0

### TypeScript version

v5.3.3

### Additional context

Please let me know what other information I can provide. I felt like I've been going down the rabbit hole, with no solution in sight..

**Comment by neehhaa06** [comment]
Hello!! @RareSecond :)

```
Ensure you're using `await` with `getCurrentUserWithRoles` in your `queryFn` and add error handling to log issues. Also, check Cypress configuration for handling server-side interactions.
```

**Comment by RareSecond** [comment]
Hi @neehhaa06 

While it shouldn't matter, I've tried your suggestion, but nothing changed.

**Comment by quintenbuis** [comment]
Having the exact same issue. With me it happen when I quickly redirect to another page but useQuery started already, leaving it in a state of always saying isLoading

**Comment by quintenbuis** [comment]
This stripped down version allowed me to continually reproduce this.

The component causing a redirect
```ts
const Checkout = (): ReactElement => {
	const { t } = useTranslation();
	const { cart, syncCart } = useCartStore();
	const [loading, setLoading] = useState<boolean>(true);

	useEffect(() => {
		if (cart && cart.lineItems.length === 0) {
			redirect('/cart');
		}

		setLoading(false);
	}, [cart]);

	if (loading) {
		return <></>;
	}

	return (
		<div className="container mb-10">
			<div className="flex flex-col lg:flex-row gap-8 mt-8">
				<PaymentMethods />
			</div>
		</div>
	);
};
```

The useQuery being used where `getPaymentMethods` is the server action:
```ts
const usePaymentMethods = (): UseQueryResult<PaymentMethod[]> => {
	const { shop } = useShopContext();

	return useQuery<PaymentMethod[]>({
		queryKey: ['paymentMethods', shop],
		queryFn: async () => {
			return await getPaymentMethods();
		},
	});
};

export default usePaymentMethods;
```

Inside PaymentMethod the useQuery is initiated due to it being rendered into the DOM, but the redirect quickly interrupts that call, causing it to be in a constant loading state.

The workaround for this is to not use server actions, but to create an API proxy yourself and there running the action. Not the most ideal fix ofcourse since it just shouldn't happen.

**Comment by adamlewkowicz** [comment]
I'm facing a similar issue. Few actions are called, but some are pending in an infinite `isPending` state, and the server action is never called.
Is there going to be any fix for this, as this is a minor issue?

**Comment by TkDodo** [comment]
> This is coming [straight from the Next.js docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#behavior)

yes, but it's still an "action". The section in the docs is called `Server Actions and Mutations`, and it's next to `Data Fetching and Caching`. Using a server action for data fetching is, again, not recommended by next or by us.

I talked to the nextjs team not too long ago and mid-term, they want to provide something like "server actions" but for fetching. But server actions ain't it.

> I want to again stress the fact that, when using a simple useEffect and (very high level, I know) recreating the behavior of react-query, everything works as expected.

I hope it's understandable that my answer here can only be:

I want to again stress the fact that, when using a simple `fetch` and (very high level, I know) recreating the behavior of server actions, everything works as expected.

Bottom line is I really don't even know how to look into this ...

**Comment by bartcheers** [comment]
> Right now, serverActions run in serial. All of them. One by one. So if you have two unrelated queries, they will be queued up.

This surprised me. I tested this and it seems that it is incorrect, unless I'm missing something - server actions can run in parallel: See this [Codesandbox](https://codesandbox.io/p/devbox/parallel-server-actions-rjj38w?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0ciho8i00063b6jd2uefb8s%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0ciho8h00023b6jde8vconn%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0ciho8h00043b6j8ida3z5h%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0ciho8i00053b6j7ro4uqf5%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0ciho8h00023b6jde8vconn%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ciho8h00013b6jhu1qg0gn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522cm0ciho8h00023b6jde8vconn%2522%252C%2522activeTabId%2522%253A%2522cm0ciho8h00013b6jhu1qg0gn%2522%257D%252C%2522cm0ciho8i00053b6j7ro4uqf5%2522%253A%257B%2522id%2522%253A%2522cm0ciho8i00053b6j7ro4uqf5%2522%252C%2522activeTabId%2522%253A%2522cm0cipb3w00373b6jvr9arrii%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cm0cipb3w00373b6jvr9arrii%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cm0ciho8h00043b6j8ida3z5h%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ciho8h00033b6jrvxoa0b1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522cm0ciho8h00043b6j8ida3z5h%2522%252C%2522activeTabId%2522%253A%2522cm0ciho8h00033b6jrvxoa0b1%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

**Comment by TkDodo** [comment]
@bartcheers your sandbox just calls an async function (that is marked as "use server") in a server component.

To really get a server action, that gets transformed to a POST request under the hood, you need to invoke the action from a client component (e.g. in a button click handler).

**Comment by bartcheers** [comment]
Thanks @TkDodo for clarifying. Here's another [Codesandbox](https://codesandbox.io/p/devbox/parallel-server-actions-forked-wn2hv9?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0ciho8i00063b6jd2uefb8s%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0ciho8h00023b6jde8vconn%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0ciho8h00043b6j8ida3z5h%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0ciho8i00053b6j7ro4uqf5%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0ciho8h00023b6jde8vconn%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ciho8h00013b6jhu1qg0gn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cm0ciho8h00023b6jde8vconn%2522%252C%2522activeTabId%2522%253A%2522cm0ciho8h00013b6jhu1qg0gn%2522%257D%252C%2522cm0ciho8i00053b6j7ro4uqf5%2522%253A%257B%2522id%2522%253A%2522cm0ciho8i00053b6j7ro4uqf5%2522%252C%2522activeTabId%2522%253A%2522cm0cipb3w00373b6jvr9arrii%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cm0cipb3w00373b6jvr9arrii%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cm0ciho8h00043b6j8ida3z5h%2522%253A%257B%2522id%2522%253A%2522cm0ciho8h00043b6j8ida3z5h%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ciho8h00033b6jrvxoa0b1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cm0ciho8h00033b6jrvxoa0b1%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) confirming server actions run in series, not in parallel. Great to know!

**Comment by RareSecond** [comment]
While I'm starting to see more why we shouldn't use it for data fetching, I do still think that there's an issue on your part.

I recreated the test in Playwright and ran it 100 times (for 3 different browsers, so 300 tests in total).

This was React Query
![image](https://github.com/user-attachments/assets/548a2c70-12d5-41cd-9152-f2f8266bf55e)

This was swr
![image](https://github.com/user-attachments/assets/6207755a-d445-4858-beb2-e4afc0d0291c)

We will be making the switch to swr, as that's an easier replacement than making the switch to API routes/tRPC.

**Comment by Dakuan** [comment]
I've hit this as well on one of my personal projects. wierdly only happening in dev. Using server actions to fetch data client side in combination with react-query and all hell breaks loose...but only with react-query

```
  const { hasNextPage, isFetchingNextPage, fetchNextPage, data, isLoading } =
    useInfiniteQuery({
      queryKey: key,
      queryFn: ({ pageParam }) => {
        const offset = (pageParam - 1) * pageSize;
        return query(offset);
      },
      refetchOnWindowFocus: false,
      initialPageParam: 1,
      getNextPageParam: (lastPage, pages) => {
        if (lastPage.length === 0) return;
        return pages.length + 1;
      },
    });
```

`lastPage` would intermittently be null 🤷

**Comment by FilipPano** [comment]
I seem to be experiencing this problem in production, but with useMutation invoking a server action:

`export function useUpdateProduct() {
  return useMutation({
    mutationFn: updateProduct, // this is a server action
    onSuccess: () => {
      toast.success('Your product has been saved.');
    },
    onError: (e) => {
      console.error(e);
      toast.error('Failed to save your product. Please try again.');
    },
  });
}`

Once this problem occurs, this and other mutations fail without even invoking the server action. Once I refresh the page they start working again.

**Comment by MarekProkopec** [comment]
Hi, I would also like to know at least some workaround for this.
I know you should not be fetching data with server actions, but the reality sadly is, that I have an entire application built on top of server action and I cannot really change that.

**Comment by maxswjeon** [comment]
Hi, just read #7970 and this issue. So should I think this issue as won't fix, or will there be a workaround available sometime else?

**Comment by dejoma** [comment]
I have an interesting case too where;
- useCustomHook1
- useCustomHook2

Both have queryFn like `queryFn: () => await someServerAction(arg1, arg2)`

One of the hooks only loads the first time, and then on the "next re-render" it just gets stuck in the loading state. QueryKeys do update, I can even see the function being triggered, but any log inside that server action does not get triggered.

**Comment by TkDodo** [comment]
It’s been almost 6 months, and no one wanted to follow through with the recommendation to update the docs. The one PR that was filed was abandoned, so my take is that it’s not important enough to be documtend.

Once again, if you come here wanting to use a server action inside a query function, please be advised that server actions are NOT RECOMMENDED to be used within query functions.

**Comment by ricoarisandyw** [comment]
Hey guys, I've ugly solution. use import.

```typescript
queryFn: async () => {
      const get = await import("@/app/actions/agent-applications")
      const res = await get.GetApplication({
        limit: 5,
        page: 1,
        search,
        status: status ? [parseInt(status)] : [],
        orderBy: sortBy,
        order: sortOrder,
        studentId,
        unread,
        latestUpdated
      })
      return res?.data
}
```

`GetApplication` is the server actions

**Comment by Walter0697** [comment]
@ricoarisandyw thanks a lot it works, but I didn't use this library at all. This seems like a nextjs issue. I am using next@15.3.0
The frustrating thing is that it doesn't happen everytime, sometimes when i route a page using either router.push or window.location.href, the target page will meet this situation.

I wish someone can point me to a nextjs issue so that we can get a prettier solution than this one

**Comment by jonator** [comment]
I'm experiencing this as well. I immediately redirect via `next/navigation` following this submit callback:

```typescript
async function onSubmit(portfolio: PortfolioForm) {
    const selectedProtocol = portfolio.protocolType
      ? protocols?.find((protocol) => protocol.type === portfolio.protocolType)
      : undefined

    setIsSubmitting(true)
    try {
      if (!selectedProtocol) {
        await createWallet({
          wallet_address: walletAddress.id,
        })
        onDone()
        return
      }

      await createWallet({
        wallet_address: walletAddress.id,
        tokens: [
          {
            type: selectedProtocol.type,
            contract_address: selectedProtocol.address,
            symbol: selectedProtocol.symbol,
          },
        ],
      })

      onDone()
    } catch (error) {
      console.error(error)
      toast.error(`Error creating wallet: ${error instanceof Error ? error.message : 'Transaction failed'}`)
    } finally {
      setIsSubmitting(false)
    }
  }
```


The createWallet function does nothing, and the redirect happens without a wallet getting created.

Next v15.3.2

**Comment by jgatto1** [comment]
I'm facing the same but without react-query! This is crazy! Any help if more than welcome!

**Comment by Greenbauer** [comment]
wrapping the action call in a timeout also seems to fix the issue 

`
queryFn: () => {
  console.log("About to get roles");
  return new Promise<UserWithRoles>((resolve, reject) => {
    setTimeout(async () => {
      try {
        const res = await getCurrentUserWithRoles();
        console.log("res: ", res);
        resolve(res);
      } catch (err) {
        reject(err);
      }
    }, 0);
  });
}
`

**Comment by Greenbauer** [comment]
wrapping the action call in a timeout also seems to fix the issue

```javascript
queryFn: () => {
  console.log("About to get roles");
  return new Promise<UserWithRoles>((resolve, reject) => {
    setTimeout(async () => {
      try {
        const res = await getCurrentUserWithRoles();
        console.log("res: ", res);
        resolve(res);
      } catch (err) {
        reject(err);
      }
    }, 0); 
  });
}
```

**Comment by clalexander** [comment]
While I'm not using React Query, I'm having the same issue in a custom hook (intermittent action calls). Settimeout worked, but I don't like it. Another workaround by starting with `Promise.resolve()` also works (I still don't like it):

```typescript
queryFn: () => {
  console.log('About to get roles');
  return Promise.resolve()
    .then(getCurrentUserWithRoles)
    .then((result) => {
      console.log('result: ', result);
      return result;
    })
    .catch((error) => {
      console.log('error: ', error);
      throw error;
    });
}
```
